lab1/myswitch_stp.py

```diff
@@ -0,0 +1,247 @@
+        if self.array[self.pos] == None:
+            self.array[self.pos] = addr
+        else:
+            evict = self.array[self.pos]
+            del self.map[evict]
+            self.array[self.pos] = addr
+
+        self.pos = (self.pos + 1) % self.size
+
+        return
+
+
+class SpanningTreeContext:
+    def __init__(self, my_id):
+        """ switches assume they're the root node upon startup
+        """
+        self.my_id = EthAddr(my_id)
+        self.root_id = EthAddr(my_id)
+        self.hops_from_root = 0
+        self.time_last_spm_tx = None
+        """ Non-root mode properties
+        """
+        # rx interface from where the STM of the perceived root came
+        self.root_interface = None
+        # ID of the switch connected to the root_interface
+        self.root_switch_id = None
+        self.blocked_interfaces = set()
+        self.time_last_spm_rx = None
+
+    def am_root(self):
+        return self.my_id == self.root_id
+
+    def become_root(self):
+        log_debug("{} becoming root".format(self.my_id))
+        self.__init__(my_id=self.my_id)
+
+    def block(self, blockee):
+        self.blocked_interfaces.add(blockee)
+
+    def unblock(self, unblockee):
+        if unblockee in self.blocked_interfaces:
+            self.blocked_interfaces.remove(unblockee)
+
+
+def emit_stm(net, interfaces, stp_context):
+    """ If am the root node and it's been 2 seconds since the last STM tx
+    then send STM out; else if am not the root and I have not seen any STM
+    packets in 10 seconds, I become the root.
+    """
+    now = datetime.now()
+    if stp_context.am_root():
+        if stp_context.time_last_spm_tx == None or \
+                (now - stp_context.time_last_spm_tx).seconds >= 2:
+
+            spm = STM.SpanningTreeMessage(root_id=stp_context.my_id,
+                                          switch_id=stp_context.my_id)
+            for intf in interfaces:
+                pkt = Ethernet(src=intf.ethaddr,
+                               dst="ff:ff:ff:ff:ff:ff",
+                               ethertype=EtherType.SLOW) + spm
+                safe_send_packet(net, intf.name, pkt)
+                log_debug("{} emitted STM on {}: {}".format(
+                    stp_context.my_id, intf.name, pkt))
+
+            stp_context.time_last_spm_tx = datetime.now()
+
+    else:
+        if stp_context.time_last_spm_rx == None or \
+                (now - stp_context.time_last_spm_rx).seconds >= 10:
+
+            stp_context.become_root()
+
+
+def handle_stm(net, interfaces, stp_context, pkt, incoming_interface):
+    stp_context.time_last_spm_rx = datetime.now()
+    log_debug("{} received STM on {}: {}".format(stp_context.my_id,
+                                                 incoming_interface, pkt))
+    stm = STM.SpanningTreeMessage()
+    b = pkt[1].to_bytes()
+    stm.from_bytes(b)
+    stm.hops_to_root += 1
+
+    def update_info_as_described_in_point_4_and_forward_STP():
```

Comment:
Yeah, agreed. I did this to purely follow his lab instructions verbatim. Also, this function is very limited in scope - it can only be called from within the `handle_stmp()' function.

To be honest, I don't have a good handle on the mechanism in which this simplified spanning tree protocol  even works as described in the lab instruction. As such, I basically just implemented the `handle_stmp()' strictly following the lab instruction. You can see the code comments are basically copy/paste of the instructions.